File: apple/RemoteDisplay/Views/Character/CharacterView.swift
//
//  CharacterView.swift
//  RemoteDisplay
//
//  Created by Júlio Flores on 17/08/2026.
//

import SwiftUI

struct CharacterView: View {
	let cell: Display.Line.Cell

	@State private var viewModel = CharacterViewModel()
	@State private var parentSize: CGSize = .zero
	@State private var textSize: CGSize = .zero

	var body: some View {
		Text(LocalizedStringResource(unicodeScalarLiteral: "\(characterOrCursor)"))
			.font(.display)
			.foregroundStyle(.character)
			.aspectRatio(5/8, contentMode: .fit)
			.background(.characterBackground)
			.onChange(of: cell, initial: true, manageBlinking)
	}

	var characterOrCursor: Character {
		switch self.cell {
			case let .char(character):
				character
			case .blank:
				" "
			case .cursor:
				self.viewModel.showCursor ? "_" : " "
		}
	}

	private func manageBlinking() {
		if case .cursor = cell {
			self.viewModel.blink()
		}
	}
}

#Preview {
	CharacterView(cell: .cursor)
		.frame(width: 50, height: 80)
		.padding()
		.background(.displayBackground)
}

File: android/app/src/main/java/dk/cipher/remotedisplay/views/character/CharacterView.kt
package dk.cipher.remotedisplay.views.character

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.text.BasicText
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.runtime.Composable
import androidx.compose.runtime.SideEffect
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import dk.cipher.remotedisplay.R
import dk.cipher.remotedisplay.models.Cell

@Composable
fun CharacterView(cell: Cell, modifier: Modifier = Modifier) {
    val viewModel: CharacterViewModel = viewModel(factory = CharacterViewModel.build())

    SideEffect(cell is Cell.Cursor) {
        if (cell is Cell.Cursor) {
            viewModel.blink()
        } else {
            viewModel.steady()
        }
    }

    val char: Char = when (cell) {
        is Cell.Character -> cell.char
        is Cell.Cursor -> if (viewModel.showCursor) '_' else ' '
        is Cell.Blank -> ' '
    }

    BasicText(
        text = "$char",
        style = TextStyle(
            color = colorResource(R.color.character),
            fontSize = 70.sp,
            fontFamily = FontFamily(Font(R.font.ninepin))
        ),
        modifier = modifier
            .aspectRatio(5f/8f, matchHeightConstraintsFirst = true)
            .background(color = colorResource(R.color.character_background))
    )
}

@Preview
@Composable
fun CharacterPreview() {
    CharacterView(
        cell = Cell.Cursor,
        modifier = Modifier
            .size(50.dp, 80.dp)
            .background(color = colorResource(R.color.display_background))
            .padding(5.dp)
    )
}